torch/csrc/jit/codegen/cuda/lower_loops.cpp

```diff
@@ -303,15 +304,234 @@ void LoopNestGenerator::handle(Expr* expr) {
+
+  TensorView* out_tv = findOutputTensor(expr);
+  if (out_tv == nullptr) {
+    target = nullptr;
+    score = 0;
+    return;
+  }
+
+  if (!out_tv->hasComputeAt()) {
+    target = out_tv;
+    // No computeAt, so this should come last.
+    score = std::numeric_limits<unsigned>::max();
+    return;
+  }
+
+  auto axis = out_tv->getRelativeComputeAtAxis();
+  target = out_tv->getComputeAtView();
+  while (target->hasComputeAt()) {
+    if (axis <= target->getThisComputeAtAxis()) {
+      // Merge this expr to the target of the target group
+      axis = target->getRelativeComputeAtAxis() -
+          (target->getThisComputeAtAxis() - axis);
+      target = target->getComputeAtView();
+    } else {
+      break;
+    }
+  }
+  score = axis;
+}
+
+// Type definitions for brevity
+using ExprListT = std::vector<Expr*>;
+using TargetGroupMapT = std::unordered_map<TensorView*, ExprListT>;
+using ExprTargetMapT = std::unordered_map<Expr*, TensorView*>;
+using ScoreT = unsigned;
+using ExprScoreMapT = std::unordered_map<const Expr*, ScoreT>;
+
+void sanityCheck(
+    const ExprListT& exprs,
+    const ExprListT& reordered_exprs,
+    const ExprScoreMapT& scores,
+    const ExprTargetMapT& target_map,
+    const TargetGroupMapT& computed_at_exprs) {
+  const auto num_exprs = exprs.size();
+  TORCH_INTERNAL_ASSERT(scores.size() == num_exprs);
+  TORCH_INTERNAL_ASSERT(
+      reordered_exprs.size() + target_map.size() == num_exprs);
+  int num_computed_exprs = std::accumulate(
+      computed_at_exprs.begin(),
+      computed_at_exprs.end(),
+      0,
+      [](int acc, const std::pair<TensorView*, ExprListT>& p) {
+        return acc + p.second.size();
+      });
+  TORCH_INTERNAL_ASSERT(num_computed_exprs == target_map.size());
+}
+
+// Arrange exprs into loop-nest groups. Loop-nest groups are
+// disjoint grouping of expressions based on the expression
+// where each expression is computed at.
+void groupExpressions(
+    Expr* expr,
+    ExprListT& reordered_exprs,
+    ExprTargetMapT& target_map,
+    TargetGroupMapT& computed_at_exprs,
+    ExprScoreMapT& scores) {
+  TensorView* target_tensor = nullptr;
+  ScoreT score;
+  findTargetTensor(expr, target_tensor, score);
+  scores.emplace(expr, score);
+  if (target_tensor == nullptr) {
+    reordered_exprs.push_back(expr);
+  } else {
+    target_map.emplace(expr, target_tensor);
+    if (computed_at_exprs.find(target_tensor) == computed_at_exprs.end()) {
+      computed_at_exprs.emplace(target_tensor, TargetGroupMapT::mapped_type());
+    }
+    auto& exprs = computed_at_exprs[target_tensor];
+    exprs.push_back(expr);
+  }
+}
+
+// Sort each loop-nest group based on axis (i.e., score)
+void sortGroup(TensorView* target, ExprListT& exprs, ExprScoreMapT& scores) {
+  std::stable_sort(
+      exprs.begin(),
+      exprs.end(),
+      [&scores](const Expr* expr1, const Expr* expr2) {
+        return scores[expr1] < scores[expr2];
+      });
+}
+
+void mergeNonRootGroupsIntoRootGroups(
```

Comment:
What's a root group?

Reply:
Having a hard time understanding what's going on in this function. Please comment.

Reply:
I added some description in the lowering section of the code gen manual. Let me know if that's not clear enough.